python3Packages.sphinxHook: Add support for multiple builders (#188032)
NixOS/nixpkgs: doc/languages-frameworks/python.section.md

with the exception of `other` (see `format` in
[`buildPythonPackage` parameters](#buildpythonpackage-parameters) for more details).

##### Using sphinxHook {#using-sphinxhook}

The `sphinxHook` is a helpful tool to build documentation and manpages
using the popular Sphinx documentation generator.
It is setup to automatically find common documentation source paths and
render them using the default `html` style.

> ```
>   outputs = [
>     "doc"
>   ];

Reviewer: ```suggestion
```nix
  outputs = [
    "out"
    "doc"
  ];
```

we should add the out output so that people do not forget it and things break.

  nativeBuildInputs = [
    sphinxHook
  ];
```

The hook will automatically build and install the artifact into the
`doc` output, if it exists. It also provides an automatic diversion
for the artifacts of the `man` builder into the `man` target.

> ```
>   outputs = [

Reviewer: ```suggestion
```nix
  outputs = [
    "out"
```

  unittestFlags = [ "-s" "tests" "-v" ];
```

##### Using sphinxHook {#using-sphinxhook}

The `sphinxHook` is a helpful tool to build documentation and manpages
using the popular Sphinx documentation generator.
It is setup to automatically find common documentation source paths and
render them using the default `html` style.

```
  outputs = [
    "out"
    "doc"
  ];

  nativeBuildInputs = [
    sphinxHook
  ];
```

The hook will automatically build and install the artifact into the
`doc` output, if it exists. It also provides an automatic diversion
for the artifacts of the `man` builder into the `man` target.

```
  outputs = [
    "out"
    "doc"
    "man"
  ];

  # Use multiple builders
  sphinxBuilders = [
    "singlehtml"
    "man"
  ];
```

> Overwrite `sphinxRoot` when the hook is unable to find your

Reviewer 2: Can we actually document how the heuristic works, e.g. `it tries: doc, docs, doc/source, docs/source and that's it!`.

Author: IMO a detail that when documented will likely drift from the actual implemenation. My recommendation is to just set the `sphinxRoot`, when it isn't found.

Reviewer 2: I am convinced. :)

- `pythonCatchConflictsHook` to check whether a Python package is not already existing.
- `pythonImportsCheckHook` to check whether importing the listed modules works.
- `pythonRemoveBinBytecode` to remove bytecode from the `/bin` folder.
- `pythonRelaxDepsHook` will relax Python dependencies restrictions for the package.
  See [example usage](#using-pythonrelaxdepshook).
- `setuptoolsBuildHook` to build a wheel using `setuptools`.
- `setuptoolsCheckHook` to run tests with `python setup.py test`.

> - `sphinxHook` to build coumentation and manages using Sphinx.

Reviewer 2: documentation, s/manages/manage? (I'm no native speaker, but I think it is infinitive form here)

Author: I meant to write manpages there.

Reviewer 2: Heh, sorry, was in "non-conscious" review mode :C

Reviewer: ```suggestion
- `sphinxHook` to build documentation and man pages using Sphinx.
```
